### PR TITLE
Bug 1339595 - Send sync engine stats to telemetry server through the Sync Ping

### PR DESCRIPTION
**Pull request by sleroux**
This patch continues the work done in https://github.com/mozilla-mobile/firefox-ios/pull/2324 and uses the collected stats to create the sync ping. To clean up the dependencies of using Telemetry across modules (Sync, Clienta) the preface of the sync ping work moves the common telemetry code into it's own framework named Telemetry and moves other target-dependent pings (ClientTelemetry, Core Ping) into their respective targets. This frees up the ability to use the same Telemetry code for our Sync ping.

The second commit takes the existing stat collecting objects and formulates a sync ping object we send to the server. I missed a couple of things in the original patch such as not passing stats along for .partial sync results so that was added here as well.

As far as the sync ping itself, right now it just sends the ping immediately and does not do any smart caching.

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
Profile.swift | 1106 | if,for,while,do statements shouldn't wrap their conditionals in parentheses. 
Profile.swift | 1147 | Function parameters should be aligned vertically if they're in multiple lines in a declaration. 
Telemetry.swift | 116 | Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. 
#### Errors

File | Line | Reason |
| --- | ----- | ----- |
Profile.swift | 657 | There should be no space before and one after any comma. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Sync/SyncTelemetry.swift [hunk 1]
```diff
@@ -148,19 +192,50 @@ extension SyncOperationStatsSession: DictionaryRepresentable {
...
 
-    public init(opStats: SyncOperationStatsSession, engineStats: [SyncEngineStatsSession]) {
+    public init(account: FirefoxAccount?, syncOperationResult: SyncOperationResult) {
         var ping: [String: Any] = [
             "version": 1,
-            "discarded": 0,
             "why": SyncPingReason.schedule.rawValue,
-            "uid": "testUID",
-            "deviceID": "testDeviceID"
+            "uid": account?.uid ?? String(repeating: "0", count: 32)
```
Comment (rnewman) [inline Sync/SyncTelemetry.swift]: Just hard-code `"00000000000000000000000000000000"` if you think this is better than making it an optional field! Less generated code.

File: Telemetry/Telemetry.swift [hunk 2]
```diff
@@ -81,6 +90,35 @@ open class Telemetry {
...
+            AppInfo.appVersion,
+            "b",
+            AppInfo.buildNumber
+        ].joined()
+        let version = ProcessInfo.processInfo.operatingSystemVersion
+        let osVersion = "\(version.majorVersion).\(version.minorVersion).\(version.patchVersion)"
+
+        return [
+            "type": type.rawValue,
+            "id": NSUUID().uuidString,
```
Comment (rnewman) [inline Telemetry/Telemetry.swift]: Just `UUID()`, surely?

File: TelemetryTests/TelemetryTests.swift [hunk 3]
```diff
@@ -1,36 +0,0 @@
-//
-//  TelemetryTests.swift
-//  TelemetryTests
-//
-//  Created by Steph Leroux on 2017-04-07.
-//  Copyright © 2017 Mozilla. All rights reserved.
```
Comment (rnewman) [inline TelemetryTests/TelemetryTests.swift]: Ah, good. Looks like this commit cleans up a bunch of stuff in the previous; do fold before landing.

File: Sync/SyncTelemetry.swift [hunk 4]
```diff
@@ -148,19 +192,50 @@ extension SyncOperationStatsSession: DictionaryRepresentable {
...
+            ping["syncs"] = [singleSync]
+        }
+
         payload = JSON(ping)
     }
+
+    private static func enginePingDataFrom(engineResults: EngineResults) -> [[String: Any]] {
+        return engineResults.map { result in
+            var engine: [String: Any] = [
+                "name": result.0
```
Comment (rnewman) [inline Sync/SyncTelemetry.swift]: Use a destructuring expression instead of `.0` and `.1`.

File: Telemetry/Telemetry.swift (hunk 2 above)
Comment (sleroux) [inline Telemetry/Telemetry.swift]: NSUUID inherits from NSObject so not entirely sure what NSUUID() would be here. Just to be safe I use the string method.

File: Sync/SyncTelemetry.swift [hunk 5]
```diff
@@ -89,14 +89,14 @@ public struct ValidationStats: Stats {
...
-    var when: Int64?
+    var when: Timestamp?
 
-    private var startUptime: Int64?
+    private var startUptime: UInt64?
 
-    public func start(when: Int64 = Int64(Date.now())) {
+    public func start(when: UInt64 = Date.now()) {
         // Convert to milliseconds as stated in the ping format
         self.when = when / 1000
```
Comment (rnewman) [inline Sync/SyncTelemetry.swift]: Wait a minute.

```
    public static func now() -> Timestamp {
        return UInt64(1000 * Date().timeIntervalSince1970)
    }

    public static func nowMicroseconds() -> MicrosecondTimestamp {
        return UInt64(1000000 * Date().timeIntervalSince1970)
    }
```

`when`, if not specified, is already milliseconds. Why are you dividing by 1000?

File: Sync/SyncTelemetry.swift [hunk 6]
```diff
@@ -89,14 +89,14 @@ public struct ValidationStats: Stats {
 
 public class StatsSession {
     var took: Int64 = 0
-    var when: Int64?
+    var when: Timestamp?
 
-    private var startUptime: Int64?
+    private var startUptime: UInt64?
```
Comment (rnewman) [inline Sync/SyncTelemetry.swift]: To clarify, name this `startUptimeNanos`.

File: Sync/SyncTelemetry.swift (hunk 5 above)
Comment (sleroux) [inline Sync/SyncTelemetry.swift]: Yup that's a problem. For some reason I though `TimeInterval` was already in milliseconds but alas it is not.

**Pull request merged**